### PR TITLE
mason publish --create-registry

### LINKED ISSUE #15079: Streamline creating local registries in Mason
From #15076 
Currently, to use a library project, which are not intended for distribution(not available on mason registry), we require : 
(1) create a local registry
(2) add that github repo to their local registry as a package
(3) include that local registry in their MASON_REGISTRY env

This issue aims to streamline the process, so that it requires a minimal number of commands from an user.

### PR DESCRIPTION
**Pull request by ankingcodes**
fixes #15079

- This PR aims to introduce a new option to publish as `mason publish --create-registry | -c path/to/local/registry` which automates the process of creating a local registry to publish our local packages.
- update tests and documentation
- made changes requested from #15986 which was closed due to mix up of commits and other conflicts.

### CODE REVIEW

**Comment by ankingcodes** [comment]
@ben-albrecht - please add a gsoc label here. Also, this is ready for review(after incorporating changes requested in #15986 .

File: tools/mason/MasonPublish.chpl [hunk 1]
```diff
@@ -78,6 +79,35 @@ proc masonPublish(ref args: list(string)) throws {
       args.append(potentialPath);
     }
 
+    if createReg {
+      var pathReg = '';
+      if args[args.size-1] == '-c' || args[args.size-1] == '--create-registry' then
```
Comment (ben-albrecht) [inline tools/mason/MasonPublish.chpl:84]: This seems fragile (e.g. if this subcommand supports a new another arg someday, this code must be updated). Why don't we just find which arg has `-c` or `--create-registry` and then assign `pathReg` to that index + 1?

File: tools/mason/MasonPublish.chpl [hunk 2]
```diff
@@ -122,6 +152,15 @@ proc masonPublish(ref args: list(string)) throws {
   }
 }
 
+/* creates a file at a given path */
+proc touch(pathToFile: string) throws {
+  const openFile = open(pathToFile, iomode.cw);
```
Comment (ben-albrecht) [inline tools/mason/MasonPublish.chpl:157]: It's common to combine the file/channel creation together:

```
  const fileWriter = open(pathToFile, iomode.cw).writer();
  const contents = "This registry was generated by mason publish --create-registry";
  fileWriter.write(contents);
  fileWriter.close();
```

File: tools/mason/MasonPublish.chpl [hunk 3]
```diff
@@ -223,7 +263,7 @@ proc publishPackage(username: string, registryPath : string, isLocal : bool) thr
  split by the MasonUtils runCommand()/
  */
 private proc commitSubProcess(dir, command) throws {
-  var spawnArgs = [command[1], command[2], command[3], command[4]];
+  var spawnArgs = [command[0], command[1], command[2], command[3], command[4]];
```
Comment (ben-albrecht) [inline tools/mason/MasonPublish.chpl:266]: How about we make this more maintainable:

```chpl
var spawnArgs = [cmd in command] cmd;
```

File: tools/mason/MasonPublish.chpl (hunk 1 above)
Comment (ankingcodes) [reply tools/mason/MasonPublish.chpl:84]: @ben-albrecht  - But `pathReg` can come before/after the option `-c` right ? How would we decide `idx+1` or `idx-1` for `pathReg` ?

Comment (ben-albrecht) [reply tools/mason/MasonPublish.chpl:84]: > But `pathReg` can come before/after the option `-c` right ? 

Good point. We could search for any positional arguments that don't start with `-` in `args[2..]` for the path and use that index. If multiple positional args are found, that's an error.

Comment (ankingcodes) [reply tools/mason/MasonPublish.chpl:84]: > If multiple positional args are found, that's an error.

But doesn't this contradict with :
 >> supports a new another arg someday,

Comment (ben-albrecht) [reply tools/mason/MasonPublish.chpl:84]: Yes 😄 . I was thinking more of flags though, e.g. `mason publish --force --create-registry foo`.

Supporting a new positional argument would be a much bigger change, unless it was simply supporting multiple registries, which wouldn't be so hard to get to from the above suggestion.

**Comment by ankingcodes** [comment]
@ben-albrecht - made changes requested. Please review and let me know for any other changes 👍🏻

File: tools/mason/MasonPublish.chpl [hunk 4]
```diff
@@ -78,6 +79,42 @@ proc masonPublish(ref args: list(string)) throws {
       args.append(potentialPath);
     }
 
+    if createReg {
+      var pathReg = '';
+      var idx_flag = 0;
+      var idx_pathReg = 0;
+      var idx = 0;
+      for arg in args {
+        if arg != "mason" && arg != "publish" {
```
Comment (ben-albrecht) [inline tools/mason/MasonPublish.chpl:88]: A loop with this structure could catch an error if a user gave multiple paths (and requires a little less manual index tracking):

```chpl
  var registryPath: string;
  for i in 2..<args.size {
    // Find positional arguments
    if !args[i].startsWith('-') {
      if registryPath == '' {
        registryPath = args[i];
      } else {
        // Multiple positional arguments is an error
        throw new owned MasonError("mason publish --create-registry expects only one path");
      }
    }
  }
  if registryPath == '' {
    // No positional arguments is an error
    throw new owned MasonError("mason publish --create-registry expects a path");
  }
```

File: tools/mason/MasonPublish.chpl [hunk 5]
```diff
@@ -263,7 +269,7 @@ proc publishPackage(username: string, registryPath : string, isLocal : bool) thr
  split by the MasonUtils runCommand()/
  */
 private proc commitSubProcess(dir, command) throws {
```
[the remark this answers is not in the document]
Comment (ankingcodes) [reply tools/mason/MasonPublish.chpl]: `command` is used as list & domains in different functions, so, either we can change lists to domains or leave it as is.

**Comment by ankingcodes** [comment]
@ben-albrecht - The problem you faced above was due to the `commitSubprocess` not working properly. For some reason, a commit was not being made on the local-registry, due to which, the copy of the local-registry at `.mason/` was not able to pull the latest version(i.e., after the package was being published).

**Comment by ben-albrecht** [comment]
@ankingcodes - I am seeing these failures:

```
[Error matching program output for mason/masonUpdateTest (execopts: 2)]
[Error matching program output for mason/chplVersion/mason-chpl-version-update (execopts: 1)]
[Error matching program output for mason/chplVersion/mason-chpl-version-update (execopts: 2)]
[Error matching program output for mason/chplVersion/mason-chpl-version-update (execopts: 15)]
[Error matching program output for mason/chplVersion/mason-chpl-version-update (execopts: 16)]
[Error matching program output for mason/chplVersion/mason-chpl-version-update (execopts: 18)]
[Error matching program output for mason/mason-example/mason-example]
[Error matching program output for mason/mason-example-with-opts/mason-example]
[Error matching program output for mason/mason-test/mason-test]
[Error matching program output for mason/publish/create-registry/create-registry]
[Summary: #Successes = 78 | #Failures = 10 | #Futures = 0 | #Warnings = 0 ]
```

**Comment by ankingcodes** [comment]
@ben-albrecht - I am not getting any of the errors in my machine. Can you give some specifics about the errors you're getting ?

**Pull request merged**